src/agentscope/service/execute_code/exec_notebook.py

```diff
@@ -0,0 +1,198 @@
+# -*- coding: utf-8 -*-
+# pylint: disable=C0301
+"""Service for executing jupyter notebooks interactively
+Partially referenced the implementation of https://github.com/geekan/MetaGPT/blob/main/metagpt/actions/di/execute_nb_code.py
+"""  # noqa
+import base64
+import uuid
+import os
+import asyncio
+from typing import Union
+from loguru import logger
+
+
+try:
+    from nbclient import NotebookClient
+    from nbclient.exceptions import CellTimeoutError, DeadKernelError
+    import nbformat
+except ImportError as import_error:
+    from agentscope.utils.tools import ImportErrorReporter
+
+    nbclient = ImportErrorReporter(import_error)
+    nbformat = ImportErrorReporter(import_error)
+    NotebookClient = ImportErrorReporter(import_error)
+
+from agentscope.service.service_status import ServiceExecStatus
+from agentscope.service.service_response import ServiceResponse
+from agentscope.file_manager import file_manager
+
+
+class NoteBookExecutor:
+    """
+    Class for executing jupyter notebooks block interactively.
+    To use the service function, you should first init the class, then call the
+    run_code_on_notebook function.
+
+    Example:
+
+        ```ipython
+        from agentscope.service.service_toolkit import *
+        from agentscope.service.execute_code.exec_notebook import *
+        nbe = NoteBookExecutor()
+        code = "print('helloworld')"
+        # calling directly
+        nbe.run_code_on_notebook(code)
+
+        >>> Executing function run_code_on_notebook with arguments:
+        >>>     code: print('helloworld')
+        >>> END
+
+        # calling with service toolkit
+        service_toolkit = ServiceToolkit()
+        service_toolkit.add(nbe.run_code_on_notebook)
+        input_obs = [{"name": "run_code_on_notebook", "arguments":{"code": code}}]
+        res_of_string_input = service_toolkit.parse_and_call_func(input_obs)
+
+        "1. Execute function run_code_on_notebook\n   [ARGUMENTS]:\n       code: print('helloworld')\n   [STATUS]: SUCCESS\n   [RESULT]: ['helloworld\\n']\n"
+
+        ```
+    """  # noqa
+
+    def __init__(
+        self,
+        notebook: nbformat.NotebookNode = nbformat.v4.new_notebook(),
+        timeout: int = 300,
+        output_dir: Union[str, None] = None,
+    ):
```

Comment:
The docstring of the constructor is missing.

Reply:
Why does the checking of `pre-commit` is invalid for the constructor function?  e.g. `-> None` is required for the constructor function in pre-commit checking.